ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java

```diff
@@ -551,31 +663,47 @@ private static TopologyTestDriver buildStreamsTopology(final Query query,
     return tests;
   }
 
+
+  private static KsqlEngine getKsqlEngine(final SchemaRegistryClient client,
+                                          final KsqlConfig ksqlConfig) {
+      final MetaStore metaStore = new MetaStoreImpl(functionRegistry);
+
+     return new KsqlEngine(
+          new FakeKafkaTopicClient(),
+          client,
+          metaStore,
+          ksqlConfig);
+  }
+
+  private static Map<String, Object> getConfigs(final Map<String, Object> additionalConfigs) {
+      final Map<String, Object> config = new HashMap<String, Object>() {{
+          put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
+          put("application.id", "KSQL-TEST");
+          put("commit.interval.ms", 0);
+          put("cache.max.bytes.buffering", 0);
+          put("auto.offset.reset", "earliest");
+          put("topology.optimization", "all");
+          put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+      }};
+      if(additionalConfigs != null){
+          config.putAll(additionalConfigs);
+      }
+      return config;
+
+  }
+
   static void shouldBuildAndExecuteQuery(final Query query) {
-    final MetaStore metaStore = new MetaStoreImpl(functionRegistry);
     final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
 
-    final Map<String, Object> config = new HashMap<String, Object>() {{
-      put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
-      put("application.id", "KSQL-TEST");
-      put("commit.interval.ms", 0);
-      put("cache.max.bytes.buffering", 0);
-      put("auto.offset.reset", "earliest");
-      put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
-    }};
+    final Map<String, Object> config = getConfigs(new HashMap<>());
     final Properties streamsProperties = new Properties();
     streamsProperties.putAll(config);
     final KsqlConfig ksqlConfig = new KsqlConfig(config);
 
-    try (final KsqlEngine ksqlEngine = new KsqlEngine(
-        new FakeKafkaTopicClient(),
-        schemaRegistryClient,
-        metaStore,
-        ksqlConfig
-    )) {
+    try (final KsqlEngine ksqlEngine = getKsqlEngine(schemaRegistryClient, ksqlConfig)) {
       query.initializeTopics(ksqlEngine);
-      final TopologyTestDriver testDriver
-          = buildStreamsTopology(query, ksqlEngine, ksqlConfig, streamsProperties);
+      final TopologyTestDriver testDriver = buildStreamsTopologyTestDriver(query, ksqlEngine, ksqlConfig, streamsProperties);
+      assertEquals(query.expectedTopology, query.generatedTopology);
```

Comment:
Comparing the strings here means that if TopologyDescription.toString changes, then this test breaks. Its also hard to fix the test if it breaks - we'd have to parse the topology description files, and they don't seem very easy to parse. Do you think it makes sense to convert the TopologyDescription to a json-serializable representation, and then write, read, and compare that?

Reply:
@rodesai I looked into converting the `TopologyDescription` into a JSON format, and it's going to take some work.  Currently using the toString representation is what we use on streams as well as any current streams users. 

What I think could be a better solution 
is to use https://zz85.github.io/kafka-streams-viz/ which is what I used when working on the topology optimization task and made finding issues in a topology

Reply:
That would still be pretty painful right? If we make a change in KafkaStreams to change the string, for example changing indentation, or printing number of partitions for topics, then we'd have to go and update every single test file. I'm fine with using the strings for now, but I still think we should switch to a more readily parsable format at some point.

Reply:
Yes, I don't disagree with you on this, I think this may take some effort and didn't want to include in this PR.  I've created a separate ticket.